spinalcordtoolbox/scripts/sct_register_to_template.py

```diff
@@ -643,6 +679,142 @@ def main(argv: Sequence[str]):
         ])
         ftmp_seg = add_suffix(ftmp_seg, '_straightAffine')
 
+        # Register spinal rootlets to template:
+        if label_type == 'rootlet':
+
```

Comment:
I am also a bit confused. I thought that this PR would use the rootlets labels to inform the registration of the anat image to the template. But after the "if" block, we continue with normal registration to template (lines 872-874).

So, is rootlets registration actually entirely separate from the normal registration? (i.e. we don't use the rootlets as `ftmp_label` in `register_wrapper` or anything?)

Maybe it will be more clear to me once this is turned into a function, since the input and output variables will be clearer.

Reply:
Maybe this figure helps: 
![figure_methods](https://github.com/user-attachments/assets/65b615d4-4a93-43c4-aace-c60fc8fe878a)

So the main difference is that the center-of-mass of the rootlets is aligned with the center-of mass of the rootlets instead of aligning the discs and a refinement is added after (steps 4 & 5) and then the last steps of the registration to template are the same than the normal pipeline (step 6), is it clearer?

`ftmp_label` here corresponds to the center-of-mass of the rootlet segmetation (that is created in the script, which uses the rootlet segmentation)

Reply:
> `ftmp_label` here corresponds to the center-of-mass of the rootlet segmetation (that is created in the script, which uses the rootlet segmentation)

Ahhh, I totally missed that `-lrootlet` gets assigned to `fname_landmarks`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/01721239c686904d3c99ded1d22759f9623077a1/spinalcordtoolbox/scripts/sct_register_to_template.py#L306-L309

Which then becomes `ftmp_label`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/01721239c686904d3c99ded1d22759f9623077a1/spinalcordtoolbox/scripts/sct_register_to_template.py#L421

That makes much more sense now. Thank you!!

Reply:
Just to clarify one more thing:

- By using `-lrootlets` for `ftmp_label`, we are using the rootlets midpoints as the landmarks for the "coarse slicewise (SI-axis) alignment" step of the usual cord-based registration to template (as you described above).
- However, in order to register the _rootlets segmentations_ to the template, we need an entirely separate registration procedure (200 lines of new registration code starting with `if label_type == 'rootlet':`)
    - The reason that this needs to be a separate step (compared to the similar "GM-informed registration") is that we do not get the benefit of a segmentation like the WM, which encodes the shape of both GM and SC.
    - And, because we essentially perform 2 different registrations, that means that we end up using two separate warping fields, too? ("rootlets -> PAM50_rootlets" and "anat -> PAM50")? 
    - Looking at the outputs, it seems like the "rootlets -> PAM50_rootlets" warping fields (`step10Warp_zmean.nii.gz`) are not output to the user, or used in regular regisration at all...

I wonder if that will be confusing to the user, as the rootlets will get registered in a kind of "opaque" way, without access to these "rootlets-specific" warping fields.

Is there no way to create a single warping field that will accurately register both the cord and the rootlets (similar to the GM-informed registration)? This would simplify the contribution greatly... :thinking:

Reply:
The rootlets registration part is only to further align the SI axis using the rootlets, but then the warping fields are concatenated to the last 2 steps of the registration, the goal of using the rootlets segmentation is only to better align the z axis, but this doesn't affect the cord at all

> And, because we essentially perform 2 different registrations, that means that we end up using two separate warping fields, too? ("rootlets -> PAM50_rootlets" and "anat -> PAM50")?

the warping field `warp_anat2template` and `warp_template2anat` will include the rootlet step

The lines under `if label_type == 'rootlet'` only replace disc alignement or vertebrae alignement step

`Is there no way to create a single warping field that will accurately register both the cord and the rootlets (similar to the GM-informed registration)? This would simplify the contribution greatly, and reduce potential variability between the two separate registrations... 🤔`

This is exactly what you get with the warp_anat2template no? 

(see line 772 and with comment)

Reply:
I totally missed the concatenation that you highlighted in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4647#discussion_r1890748283.

My bad!! Thank you so much for explaining.